### PR TITLE
pkp/pkp-lib#6257 Apply additional policies to submission file uploads

### PR DESCRIPTION
**Pull request by NateWr**

### CODE REVIEW

File: classes/security/authorization/NoteAccessPolicy.inc.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+		$noteDao = DAORegistry::getDAO('NoteDAO'); /* @var $noteDao NoteDAO */
+		$note = $noteDao->getById($this->_noteId);
+
+		if (!is_a($note, 'Note')) {
+			return AUTHORIZATION_DENY;
+		}
+
+		// Note, query, submission and assigned stages must match
+		if ($note->getAssocId() != $query->getId()
```
Comment (asmecher) [inline classes/security/authorization/NoteAccessPolicy.inc.php:75]: Wherever `assoc_id` is checked against an ID, the `assoc_type` should also be checked to ensure it matches the right `ASSOC_TYPE_`... constant. (Applies twice in this `if`.)

File: classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	function dataObjectEffect() {
+		AppLocale::requireComponents([LOCALE_COMPONENT_PKP_SUBMISSION, LOCALE_COMPONENT_APP_SUBMISSION]);
+
+		$assignedFileStageIds = $this->getAuthorizedContextObject(ASSOC_TYPE_ACCESSIBLE_FILE_STAGES);
+		if (empty($assignedFileStageIds) || !in_array(SUBMISSION_FILE_PROOF, $assignedFileStageIds)) {
+			return AUTHORIZATION_DENY;
+		}
+
+		if (empty($this->_representationId))  {
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'user.authorization.representationNotFound');
```
Comment (asmecher) [inline classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php:52]: Are references to `$msg` left-over debugging code?

File: classes/submission/SubmissionFileDAO.inc.php [hunk 3]
```diff
@@ -781,6 +783,89 @@ public function getWorkflowStageId($submissionFile) {
...
+			if ($action === SUBMISSION_FILE_ACCESS_READ || $hasEditorialAssignment) {
+				$allowedFileStageIds[] = SUBMISSION_FILE_REVIEW_REVISION;
+			}
+			// Authors can never access review files
+			if ($hasEditorialAssignment) {
+				$allowedFileStageIds[] = SUBMISSION_FILE_REVIEW_FILE;
+			}
+		}
+
+		HookRegistry::call('SubmissionFile::assignedFileStageIds', [&$allowedFileStageIds, $stageAssignments, $action]);
```
Comment (asmecher) [inline classes/submission/SubmissionFileDAO.inc.php:864]: Should the hook name be `SubmissionFileDAO`?

File: classes/submission/SubmissionFileDAO.inc.php [hunk 4]
```diff
@@ -781,6 +783,89 @@ public function getWorkflowStageId($submissionFile) {
...
+	 * authors are granted write access to submission files or revisions only when other
+	 * conditions are met. This method does not return those as allowed file
+	 * stages for author assignments.
+	 *
+	 * @param array $stageAssignments The stage assignments of this user.
+	 *   Each key is a workflow stage and value is an array of assigned roles
+	 * @param int $action Read or write to file stages. One of SUBMISSION_FILE_ACCESS_
+	 * @return array List of file stages (SUBMISSION_FILE_*)
+	 */
+	public function getAssignedFileStageIds($stageAssignments, $action) {
```
Comment (asmecher) [inline classes/submission/SubmissionFileDAO.inc.php:807]: This feels more like a service class function than a DAO function to me (or in much older -- and very deprecated -- terms, an Action class). I'm OK to leave it here but wouldn't want this kind of code to get too entrenched in DAOs without broader consideration.

File: controllers/wizard/fileUpload/PKPFileUploadWizardHandler.inc.php [hunk 5]
```diff
@@ -63,6 +66,116 @@ function __construct() {
 	//
 	// Implement template methods from PKPHandler
 	//
+	function authorize($request, &$args, $roleAssignments) {
```
Comment (asmecher) [inline controllers/wizard/fileUpload/PKPFileUploadWizardHandler.inc.php:69]: This kind of reiteration of rules also shouldn't live here, but the upload wizard already has that kind of pollution and IMO we'll need a 3rd-party implementation to eventually reconcile all of this into.

File: controllers/wizard/fileUpload/PKPFileUploadWizardHandler.inc.php [hunk 6]
```diff
@@ -408,7 +522,12 @@ function confirmRevision($args, $request) {
 		// FIXME?: need assocType and assocId? Not sure if they would be used, so not adding now.
 		$reviewRound = $this->getReviewRound();
 		$confirmationForm = new SubmissionFilesUploadConfirmationForm(
```
Comment (asmecher) [inline controllers/wizard/fileUpload/PKPFileUploadWizardHandler.inc.php:524]: (I'd also like to come up with a better solution to function parameter lists that grow like this. We have some ideas for the DAOs and the service classes solve some problems using `$arg` arrays, but introduce others at the same time. I'm sure there are some clever PHP7 tools that are in adoption in other projects already.)

File: classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php (hunk 2 above)
Comment (NateWr) [reply classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php:52]: This avoids a pass-by-reference notice or something like that. I can't remember the exact PHP notice, but it occurs when passing a string in directly to `setAdvice`.

File: classes/submission/SubmissionFileDAO.inc.php (hunk 3 above)
Comment (NateWr) [reply classes/submission/SubmissionFileDAO.inc.php:864]: Ah, yeah, hmm. So this is backporting from a method that exists in `SubmissionFileService` in the `master` branch. So it won't live in `SubmissionFileDAO` for long. Probably better to keep the hook name from changing, even if it's a little strange appearing in the `SubmissionFileDAO` for the stable branches. Does that sound ok?

File: classes/submission/SubmissionFileDAO.inc.php (hunk 4 above)
Comment (NateWr) [reply classes/submission/SubmissionFileDAO.inc.php:807]: Yep! It is a service class method in `master`. We just don't have `SubmissionFileService` in the stable branches.

File: controllers/wizard/fileUpload/PKPFileUploadWizardHandler.inc.php (hunk 5 above)
Comment (NateWr) [reply controllers/wizard/fileUpload/PKPFileUploadWizardHandler.inc.php:69]: Agreed. One of the things that I've come to feel is that `fileStage` is maybe not appropriate for review attachments, discussion files, and dependent files. Authorization to read/write these files is done independently of their file stage. I've got some notes down to discuss this once the current work is all done and dusted.

File: controllers/wizard/fileUpload/PKPFileUploadWizardHandler.inc.php (hunk 6 above)
Comment (NateWr) [reply controllers/wizard/fileUpload/PKPFileUploadWizardHandler.inc.php:524]: Agreed. The `$args` arrays are good in that they are extensible, which makes it easy to work across applications and plugins. But they're terrible for defining expected variable names/types, and autocomplete in IDEs. I think a lot of this is improved with more careful code design, but I think that one approach is to use DependencyInjection. That's what Pimple is for (which we use under the hood for Service classes), but I've never really got my head around it.

File: classes/security/authorization/NoteAccessPolicy.inc.php (hunk 1 above)
Comment (NateWr) [reply classes/security/authorization/NoteAccessPolicy.inc.php:75]: Thanks! Fixed across all PRs.

File: classes/submission/SubmissionFileDAO.inc.php (hunk 3 above)
Comment (asmecher) [reply classes/submission/SubmissionFileDAO.inc.php:864]: Yes, that's OK with me.

File: classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php (hunk 2 above)
Comment (asmecher) [reply classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php:52]: Ah, gotcha. I just reviewed the use cases for `setAdvice` and there's no need for a reference to be used; I'd recommend getting rid of it (along with the `$msg`).

Comment (NateWr) [reply classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php:52]: Get rid of `setAdvice`?

Comment (asmecher) [reply classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php:52]: Sorry, I wasn't clear -- no, just the reference on the second parameter.

Comment (NateWr) [reply classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php:52]: So it isn't a warning but a fatal error. Just to make sure we understand each other, here's the change I made:

```diff
diff --git a/classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php b/classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php
index 76e3a2ad5..30b3c47e8 100644
--- a/classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php
+++ b/classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php
@@ -49,7 +49,7 @@ class RepresentationUploadAccessPolicy extends DataObjectRequiredPolicy {
 		}
 
 		if (empty($this->_representationId))  {
-			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'user.authorization.representationNotFound');
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, 'user.authorization.representationNotFound');
 			return AUTHORIZATION_DENY;
 		}
 
@@ -57,31 +57,31 @@ class RepresentationUploadAccessPolicy extends DataObjectRequiredPolicy {
 		$representation = $representationDao->getById($this->_representationId);
 
 		if (!$representation) {
-			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'user.authorization.representationNotFound');
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, 'user.authorization.representationNotFound');
 			return AUTHORIZATION_DENY;
 		}
 
 		$submission = $this->getAuthorizedContextObject(ASSOC_TYPE_SUBMISSION);
 		if (!$submission) {
-			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'user.authorization.invalidSubmission');
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, 'user.authorization.invalidSubmission');
 			return AUTHORIZATION_DENY;
 		}
 
 		$publication = Services::get('publication')->get($representation->getData('publicationId'));
 		if (!$publication) {
-			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'galley.publicationNotFound');
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, 'galley.publicationNotFound');
 			return AUTHORIZATION_DENY;
 		}
 
 		// Publication and submission must match
 		if ($publication->getData('submissionId') !== $submission->getId()) {
-			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'user.authorization.invalidPublication');
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, 'user.authorization.invalidPublication');
 			return AUTHORIZATION_DENY;
 		}
 
 		// Representations can not be modified on published publications
 		if ($publication->getData('status') === STATUS_PUBLISHED) {
-			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'galley.editPublishedDisabled');
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, 'galley.editPublishedDisabled');
 			return AUTHORIZATION_DENY;
 		}
 ```

And here's the fatal error it produces:

```
[15-Oct-2020 13:42:43 Europe/London] PHP Fatal error:  Uncaught Error: Cannot pass parameter 2 by reference in /home/nate/Projects/pkp/ojs-stable/lib/pkp/classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php:60
Stack trace:
#0 /home/nate/Projects/pkp/ojs-stable/lib/pkp/classes/security/authorization/DataObjectRequiredPolicy.inc.php(77): RepresentationUploadAccessPolicy->dataObjectEffect()
#1 /home/nate/Projects/pkp/ojs-stable/lib/pkp/classes/security/authorization/AuthorizationDecisionManager.inc.php(196): DataObjectRequiredPolicy->effect()
#2 /home/nate/Projects/pkp/ojs-stable/lib/pkp/classes/security/authorization/AuthorizationDecisionManager.inc.php(125): AuthorizationDecisionManager->_decidePolicySet(Object(PolicySet), NULL)
#3 /home/nate/Projects/pkp/ojs-stable/lib/pkp/classes/handler/PKPHandler.inc.php(299): AuthorizationDecisionManager->decide()
#4 /home/nate/Projects/pkp/ojs-stable/lib/pkp/controllers/wizard/fileUpload/PKPFileUploadWizardHandler.inc.php(176): PKPHandler->authorize(Object(Request), Array, Array)
#5 /home/nate/Pro in /home/nate/Projects/pkp/ojs-stable/lib/pkp/classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php on line 60
```

This is with PHP 7.4.10.

Comment (asmecher) [reply classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php:52]: You can also remove the reference from the `setAdvice` function: https://github.com/pkp/pkp-lib/blob/master/classes/security/authorization/AuthorizationPolicy.inc.php#L58

Comment (NateWr) [reply classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php]: Oh, yeah, that makes sense. Ok, I've updated all of the PRs.

**Pull request merged**